src/common/resources.cpp

```diff
@@ -1475,6 +1475,20 @@ ostream& operator<<(ostream& stream, const Volume& volume)
   return stream;
 }
 
+ostream& operator<<(ostream& stream, const Resource::DiskInfo::Source& source)
+{
+  stream << source.type();
+
+  if (source.has_path()) {
```

Comment:
It looks like the common pattern is to switch on the type. Could you please follow this pattern? You can convert `has_*()` into `CHECK`s.